content/lib/2021-04-02-operating-system-interface.md

```diff
@@ -0,0 +1,225 @@
+The first step in progress towards this goal, is defining an interface.
+
+The Distributed Operating System Interface (`DOSi`) hopes to achieve this.
+
+# Problem Statement 
+
+Over the past 7 years Kubernetes has grown in popularity. 
+Kubernetes is built and managed on operating systems optimized for isolated workloads.
+Kubernetes mutates operating system level features, concerns, and services.
+Some of these services (CNI, Kubelet, Container Runtime, etc) are paramount for Kubernetes to run.
+
+I believe there should be a clear communication and management interface between Kubernetes and the underlying operating system.
+
+This interface does not exist, therefore an operating system implementation does not exist.
+
+# Vocabulary 
+
+#### Distributed Operating System Interface
+
+(_DOSi_) the implementable boundary that is defined and optimized for higher level interaction with the operating system.
+
+Based on the definitions below, `DOSi` is the interface between `kubernetesland` and `userland`.
+
+ - Alternatives (DOS, DOS API, DOSIX)
+
+### Kernelland 
+
+The computational space that exists below the system call API.
+
+### Userland
+
+Also known as `userspace` in the Linux community, this is the operating system layer in which humans can engage. 
+
+Everything above the kernel, that is not managed with Kubernetes.
+
+### Kubernetesland
+
+Everything above `userland` that is managed with Kubernetes
+
+### Node
+
+A kubernetes node. A single instance of a single operating system managing a machine.
+A node is a computer, that has a kubernetes-like layer running on top of the operating system.
+
+#### The Stack
+
+The combined layers of `kubernetesland`, `userland`, and `kerneland`
+
+# Summary
+
+I hope to define a clear interface between `kubernetesland` and `userland` that is complimentary to existing interfaces (CNI, CRI, CSI, OCI).
+While also considering the clear void that the sum of these existing interfaces do not address. 
+In particular management, and security of the operating system and it's components as a whole.
+
+
+Below, find benefits of defining this interface. 
+
+
+### Benefits
+
+ - Implementation, control, and management of the non-existent `kubernetes` and `node` security boundary.
+ - Clearly defined management of `node` level services and configuration.
+ - Finite set of use cases, and capabilities required from an operating system to run Kubernetes.
+    - This will define the touch points between an operating system and Kubernetes, thus outlining which operating system features are critical.
+    - EX: Do we need POSIX compliant user/group management? Or will Linux capabilities suffice?
+ - Management of required `node` level services can be safely controlled from `kubernetesland`
+    - EX: CNI mutations
+    - EX: Privileged DaemonSets for mutating a host
+    - EX: Kernel security controls (Seccomp, SELinux, eBPF, Auth, User Management)
+    - EX: Storage providers interacting with the host
+    
+# Prior Art
```

Comment:
this is the part that scares me the most - i am far from an expert in a lot of these very complex and mature solutions

Reply:
if anyone reading this has more knowledge here please do enlighten me and correct me as needed